extensions/data-plane/data-plane-api/src/main/java/org/eclipse/edc/connector/dataplane/api/controller/DataPlanePublicApiController.java

```diff
@@ -149,7 +150,7 @@ private void handle(ContainerRequestContext context, AsyncResponse response) {
                         if (result.succeeded()) {
                             response.resume(Response.ok(stream.toString()).build());
                         } else {
-                            response.resume(internalErrors(result.getFailureMessages()));
+                            response.resume(badGatewayErrors(result.getFailureMessages()));
```

Comment:
please add a test to verify this condition

Reply:
plus, I'm not sure that only "bad gateway" errors will finish here, in the `ParallelSink` and in the `OutputStreamDataSink` there are multiple reasons why an `Exception` could be thrown and caught and transformed into a `StateResult.failure`

Reply:
To step back, is the issue being addressed error correlation (see my comment below), or that 502 should be returned instead of 500?

Reply:
I agree with that, although that is the way both DataSink implementations (and their subclasses) deal with it, they wrap up the entire transfer method code into one single try/catch block which catches a generic Exception and treats anything that may go wrong down da stack as an internal error. Considering that the methods down the stack mostly rely on external dependencies (DataSource and DataSink) I figured it would make sense to treat errors thrown as a bad gateway. In order to improve error handling it would be necessary to change the DataSink and DataSource openPartStream and openStream implementations identifying and wrapping code blocks addressing accurately error causes throwing proper exceptions, because most of them (if not all of them) would also wrap the entire code block catching a generic exception. Would you guys think that is the case?

Reply:
Back-end exceptions should not be available to clients as this information is internal and potentially sensitive. The failure messages can provide correlation information but should not pass exception details. 

For me, 500 is the correct response because the fact that one data plane may be a proxy is an internal detail. For example, another data plane may not be a proxy and pull data from some storage device.

Reply:
> Back-end exceptions should not be available to clients as this information is internal and potentially sensitive. The failure messages can provide correlation information but should not pass exception details.
> 
- I agree.

> For me, 500 is the correct response because the fact that one data plane may be a proxy is an internal detail. For example, another data plane may not be a proxy and pull data from some storage device.
> 
- Then in the cases when it plays a proxy role, a failure caused by external dependencies shall be treated as internal, maybe to be more embracing I'd say, I got that.

> `DataSource` and `DataSink` should throw unchecked exceptions only to indicate non-recoverable conditions. If there is a recoverable condition, the implementations should deal with it.
> 
- That for sure, is the desired behavior, but the `DataSouce` and `DataSink` implementations, as mentioned above, only catch a generic `Exception` (which includes checked and unchecked) indistinctly making any condition irrecoverable.

Reply:
> * That for sure, is the desired behavior, but the `DataSouce` and `DataSink` implementations, as mentioned above, only catch a generic `Exception` (which includes checked and unchecked) indistinctly making any condition irrecoverable.

I don't follow. The implementations correctly handle exceptions and convert unrecoverable ones into `EdcException`. What is the issue? Note that a checked exception can be unrecoverable, e.g. `FileNotFoundException` as in `FileTransferDataSource`.

Reply:
e.g `HttpDataSource` turns any non-2xx status returned into an `EdcException` which is eventually caught as an `Exception` in `DataSink `implementations which despises the response content.